src/database_migration.py

```diff
@@ -0,0 +1,125 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Provide the DatabaseMigration class to manage database migrations."""
+
+import logging
+from typing import Literal, cast
+
+import ops
+from ops.pebble import ExecError
+
+from charm_state import CharmState
+from constants import FLASK_STATE_DIR
+from exceptions import CharmConfigInvalidError
+
+logger = logging.getLogger(__name__)
+
+
+DatabaseMigrationStatus = Literal["PENDING", "COMPLETED", "FAILED"]
+
+
+class DatabaseMigration:
+    """The DatabaseMigration class that manages database migrations.
+
+    Attrs:
+        script: the database migration script.
+        COMPLETED: one of the possible database migration status.
+        FAILED: one of the possible database migration status.
+        PENDING: one of the possible database migration status.
+    """
+
+    _STATUS_FILE = FLASK_STATE_DIR / "database-migration-status"
+    _COMPLETED_SCRIPT_FILE = FLASK_STATE_DIR / "completed-database-migration"
+
+    COMPLETED: DatabaseMigrationStatus = "COMPLETED"
+    FAILED: DatabaseMigrationStatus = "FAILED"
+    PENDING: DatabaseMigrationStatus = "PENDING"
+
+    def __init__(self, flask_container: ops.Container, charm_state: CharmState):
+        """Initialize the DatabaseMigration instance.
+
+        Args:
+            flask_container: The flask container object.
+            charm_state: The charm state.
+        """
+        self._container = flask_container
+        self._charm_state = charm_state
+
+    @property
+    def script(self) -> str | None:
+        """Get the database migration script."""
+        return self._charm_state.database_migration_script
+
+    def get_status(self) -> DatabaseMigrationStatus:
+        """Get the database migration run status.
+
+        Returns:
+            One of "PENDING", "COMPLETED", or "FAILED".
+        """
+        return cast(
+            DatabaseMigrationStatus,
+            self.PENDING
+            if not self._container.exists(self._STATUS_FILE)
+            else self._container.pull(self._STATUS_FILE).read(),
+        )
+
+    def _set_status(self, status: DatabaseMigrationStatus) -> None:
+        """Set the database migration run status.
+
+        Args:
+            status: One of "PENDING", "COMPLETED", or "FAILED".
+        """
+        self._container.make_dir(self._STATUS_FILE.parent, make_parents=True)
+        self._container.push(self._STATUS_FILE, source=status)
```

Comment:
container.push has a make_dirs=True parameter. Would that work here instead of make_dir?